## THIS IS THE CODE FOR Ambition Jewels consider PO: NL PO# 101536 AMBITION (SHIMAYRA)-EDITING.pdf

### same code for retriving data from pdf as you sent obu:

In [2]:
import pandas as pd  # #PDF TO TEXT PO1
import re
import pdfplumber

def extract_data_from_pdf(pdf_path):
    """
    Extract data from PDF file and return structured data
    """
    data = []
    
    try:
        with pdfplumber.open(pdf_path) as pdf:
            # Extract text from all pages
            full_text = ""
            for page in pdf.pages:
                full_text += page.extract_text() + "\n"
            
            print("Extracted text from PDF:")
            print("=" * 50)
            print(full_text)
            print("=" * 50)
            
            # Process the extracted text
            lines = full_text.split('\n')
            current_item = {}
            
            for line in lines:
                line = line.strip()
                if not line:
                    continue
                    
                # Extract SrNo (looking for lines starting with numbers)
                sr_no_match = re.match(r'^\s*(\d+)\s*', line)
                if sr_no_match:
                    if current_item and 'SrNo' in current_item:  # Save previous item if exists
                        data.append(current_item)
                    current_item = {'SrNo': sr_no_match.group(1), 'OrderQty': 2}
                    
                # Extract Article code (patterns like 9-DD106-YG-25, 9-DD106-YG-7)
                article_match = re.search(r'([A-Z0-9\-]+YG\-?\d*\.?\d+)', line)
                if article_match and 'ArticleCode' not in current_item:
                    current_item['ArticleCode'] = article_match.group(1)
                    
                # Extract StyleCode from Your reference (text after RSBR2074-01, RSBR2074-03, etc.)
                style_match = re.search(r'RSBR2074\-(\d+)\s+([A-Z0-9]+)', line)
                if style_match:
                    current_item['StyleCode'] = style_match.group(2)
                    
                # Extract ItemSize from description (looking for numbers like 0.25, 0.07)
                # Multiple patterns to catch different formats
                size_patterns = [
                    r'pendant\s+(\d+\.\d+)',  # pendant 0.25
                    r'(\d+\.\d+)\s*ct',       # 0.25 ct
                    r'YG[-\s]*(\d+\.\d+)',    # YG-0.25 or YG 0.25
                ]
                
                for pattern in size_patterns:
                    size_match = re.search(pattern, line, re.IGNORECASE)
                    if size_match and 'ItemSize' not in current_item:
                        current_item['ItemSize'] = size_match.group(1)
                        break
            
            # Add the last item
            if current_item and 'SrNo' in current_item:
                data.append(current_item)
                
    except Exception as e:
        print(f"Error reading PDF file: {e}")
        return []
    
    return data

def create_excel_dataframe(data):
    """
    Create DataFrame with specified columns
    """
    if not data:
        print("No data extracted from PDF")
        return pd.DataFrame()
    
    # Create DataFrame
    df = pd.DataFrame(data)
    
    # Reorder columns as requested
    columns_order = ['SrNo', 'StyleCode', 'ItemSize', 'OrderQty']
    
    # Only include columns that exist in the data
    final_columns = [col for col in columns_order if col in df.columns]
    
    # Add missing columns with empty values
    for col in columns_order:
        if col not in df.columns:
            df[col] = ""
    
    return df[columns_order]

# Main execution
if __name__ == "__main__":
    # Specify your PDF file path here
   
    pdf_file_path = r'C:\Users\Pratik Mali\Desktop\tools\data\Ambition Jewels Pvt. Ltd (AJP)\NL PO# 101536 AMBITION (SHIMAYRA)-EDITING.pdf'  # Change this to your actual PDF path
    
    # Extract data from PDF
    print(f"Reading PDF from: {pdf_file_path}")
    extracted_data = extract_data_from_pdf(pdf_file_path)
    
    if extracted_data:
        # Create DataFrame
        df = create_excel_dataframe(extracted_data)
        
        # Display the results
        print("\nExtracted Data:")
        print("=" * 50)
        print(df)
        
        # Save to Excel file
        output_file = 'extracted_data.xlsx'
        df.to_excel(output_file, index=False)
        print(f"\nData successfully saved to '{output_file}'")
        
        # Display basic statistics
        print(f"\nExtraction Summary:")
        print(f"Total records extracted: {len(df)}")
        print(f"Columns: {list(df.columns)}")
        
    else:
        print("No data was extracted from the PDF file.")
    
    # Display the DataFrame in notebook
    df

Reading PDF from: C:\Users\Pratik Mali\Desktop\tools\data\Ambition Jewels Pvt. Ltd (AJP)\NL PO# 101536 AMBITION (SHIMAYRA)-EDITING.pdf
Extracted text from PDF:
P URCHASE ORDER
PO # : 101536
Date : Sep/24/2025
# Order # SO # Style # Description Metal Color Size Order Price Finished Diamond Due Vendor Workbag #
Type Qty Wt Quality Date Style
1. 101536/101 E14639 LN2095SS-24 1/2 CT LCD BUTTERCUP SILV W 18.00 38 10.00 *LG3 Nov/12/2025 ZNLSN00814 3079860000
ETERNIT NECK GSI CB
Special Inst. STAMP 925 LGD ANJ, DIA I SI2, GROSS WT. 10.20 GMS. **GSI MIN CERT** **NO GSI# ENGRAVING
ON PCS** **DIRECT SHIP TO CANADA**
Comments SILVER **FOR SHIMAYRA**
Total : 38
Page 1 of 1


Extracted Data:
  SrNo StyleCode ItemSize  OrderQty
0    1                            2

Data successfully saved to 'extracted_data.xlsx'

Extraction Summary:
Total records extracted: 1
Columns: ['SrNo', 'StyleCode', 'ItemSize', 'OrderQty']


In [3]:
import pandas as pd  # #PDF TO TEXT PO1
import re
import pdfplumber

def extract_data_from_pdf(pdf_path):
    """
    Extract data from PDF file and return structured data
    """
    data = []
    
    try:
        with pdfplumber.open(pdf_path) as pdf:
            # Extract text from all pages
            full_text = ""
            for page in pdf.pages:
                full_text += page.extract_text() + "\n"
            
            print("Extracted text from PDF:")
            print("=" * 50)
            print(full_text)
            print("=" * 50)
            
            # Process the extracted text
            lines = full_text.split('\n')
            current_item = {}
            
            for line in lines:
                line = line.strip()
                if not line:
                    continue
                    
                # Extract SrNo (looking for lines starting with numbers)
                sr_no_match = re.match(r'^\s*(\d+)\s*', line)
                if sr_no_match:
                    if current_item and 'SrNo' in current_item:  # Save previous item if exists
                        data.append(current_item)
                    current_item = {'SrNo': sr_no_match.group(1), 'OrderQty': 2}
                    
                # Extract Article code (patterns like 9-DD106-YG-25, 9-DD106-YG-7)
                article_match = re.search(r'([A-Z0-9\-]+YG\-?\d*\.?\d+)', line)
                if article_match and 'ArticleCode' not in current_item:
                    current_item['ArticleCode'] = article_match.group(1)
                    
                # Extract StyleCode from Your reference (text after RSBR2074-01, RSBR2074-03, etc.)
                style_match = re.search(r'RSBR2074\-(\d+)\s+([A-Z0-9]+)', line)
                if style_match:
                    current_item['StyleCode'] = style_match.group(2)
                    
                # Extract ItemSize from description (looking for numbers like 0.25, 0.07)
                # Multiple patterns to catch different formats
                size_patterns = [
                    r'pendant\s+(\d+\.\d+)',  # pendant 0.25
                    r'(\d+\.\d+)\s*ct',       # 0.25 ct
                    r'YG[-\s]*(\d+\.\d+)',    # YG-0.25 or YG 0.25
                ]
                
                for pattern in size_patterns:
                    size_match = re.search(pattern, line, re.IGNORECASE)
                    if size_match and 'ItemSize' not in current_item:
                        current_item['ItemSize'] = size_match.group(1)
                        break
            
            # Add the last item
            if current_item and 'SrNo' in current_item:
                data.append(current_item)
                
    except Exception as e:
        print(f"Error reading PDF file: {e}")
        return []
    
    return data

def create_excel_dataframe(data):
    """
    Create DataFrame with specified columns
    """
    if not data:
        print("No data extracted from PDF")
        return pd.DataFrame()
    
    # Create DataFrame
    df = pd.DataFrame(data)
    
    # Reorder columns as requested
    columns_order = ['SrNo', 'StyleCode', 'ItemSize', 'OrderQty']
    
    # Only include columns that exist in the data
    final_columns = [col for col in columns_order if col in df.columns]
    
    # Add missing columns with empty values
    for col in columns_order:
        if col not in df.columns:
            df[col] = ""
    
    return df[columns_order]

# Main execution
if __name__ == "__main__":
    # Specify your PDF file path here
   
    pdf_file_path = r'C:\Users\Pratik Mali\Desktop\tools\data\Ambition Jewels Pvt. Ltd (AJP)\NL PO# 101533 AMBITION (SHIMAYRA)-EDITING.pdf'  # Change this to your actual PDF path
    
    # Extract data from PDF
    print(f"Reading PDF from: {pdf_file_path}")
    extracted_data = extract_data_from_pdf(pdf_file_path)
    
    if extracted_data:
        # Create DataFrame
        df = create_excel_dataframe(extracted_data)
        
        # Display the results
        print("\nExtracted Data:")
        print("=" * 50)
        print(df)
        
        # Save to Excel file
        output_file = 'extracted_data.xlsx'
        df.to_excel(output_file, index=False)
        print(f"\nData successfully saved to '{output_file}'")
        
        # Display basic statistics
        print(f"\nExtraction Summary:")
        print(f"Total records extracted: {len(df)}")
        print(f"Columns: {list(df.columns)}")
        
    else:
        print("No data was extracted from the PDF file.")
    
    # Display the DataFrame in notebook
    df

Reading PDF from: C:\Users\Pratik Mali\Desktop\tools\data\Ambition Jewels Pvt. Ltd (AJP)\NL PO# 101533 AMBITION (SHIMAYRA)-EDITING.pdf
Extracted text from PDF:
P URCHASE ORDER
PO # : 101533
Date : Sep/24/2025
# Order # SO # Style # Description Metal Color Size Order Price Finished Diamond Due Vendor Workbag #
Type Qty Wt Quality Date Style
1. 101533/101 E14637 LB2354SS-24 2 CTW LCD RD TENNIS SILV W 7.00 30 7.25 *LG3 Nov/12/2025 BR0000094K 3079670000
BRACELET GSI CB
Special Inst. STAMP 925 LGD ANJ, DIA I SI2, GROSS WT (WITH TAG) : 7.77 GMS, 3K TONGUE, **DIRECT SHIP TO
CANADA** **GSI MIN CERT** **NO GSI# ENGRAVING ON PCS** ** FOR SHIMAYRA**
Comments SILVER **FOR SHIMAYRA**
2. 101533/102 E14637 LB2353SSYV-24 1 CTW LCD DIA RD TENNIS SILV YV 7.00 75 6.60 *LG3 Nov/12/2025 BR0000081H 3079680000
BRACELET GSI CB
Special Inst. STAMP 925 10KGP LGD ANJ, DIA I SI2, GROSS WT (WITH TAG) : 6.90 GMS, **GSI MIN CERT NO
GSI# ENGRAVING ON PCS** WHITE RHODIUN ON TIP PRONG. (10K 2.5 MICRON YELLOW
PLATING), 

#logic for the AMBITION PO:

**PROMPT**:

consider the first cell in the @Ambition.ipynb file from which the text is extracted for the text extracted create a dataframe and arrange that text in the following columns in a dataframe and save that in a excel sheet, so add the cell accordingly at the end of the @Ambition.ipynb file to do so, the instructions for each columns to be filled are as follows,

COLUMNS:

for each item the details will start from the Sr.No. here the first item starts from "1." in the text and the data for that item lasts till  "**FOR SHIMAYRA**" is seen
1.SrNo: The first number that is present after the 5th line in the text for example here after the 5th line after the word "style" in next line "1." is the Sr.No. for this item similarly for the next ITem if at the end of the item after "**FOR SHIMAYRA**" is next Sr.No. if is not seen it then means that the file end there do not consider any thing after that.

2.StyleCode: the StyleCode for a item is the second last code in the Sr.No. line for example here the style code for first item is: 'ZNLSN00814' similarly find for other items

3.ItemSize:itemsize is the first decimal number that is present in the Sr.No. line for example here item size is "18.00"

4.OrderQty: the number that is present right after the ItemSize is the orderQty for example here order quantity is "38"

5.OrderItemPcs: leave this column blank

6.Metal: for different items there are difference Metal code for example for example if "SILV" is present in the text then it is AG925 so whenever "SILV" is encountered assign "AG925" as Metal for that item and similary "G14W" for Gold 14kt present in the text and "W" is the tone present which will depend on the tone column 

7.Tone: after the "SILV" word/letter is encountered for that item there will be the tone for example here "W" is the tone

8.ItemPoNo.: its the same for all the items here in the text at the top PO # : 101533 "101533" is the ItemPoNo

9.ItemRefNo: the first digits that are present just after the Sr.No. is the ItemRefNo for example here "101536/101" is the ItemRefNo for that item

10.StockType: leave it blank

11.MakeType: leave it blank

12.CustomerProductionInstruction: for example here "1/2 CT LCD BUTTERCUP ETERNIT NECK GSI CB" is the CustomerProductionInstruction for that item similarly find for other items too if present

13.SpecialRemarks: here "101536/101  LN2095SS-24 SILV  SZ-18inch,  DIA I SI2" is the SpecialRemarks for the item so find similarly where it is present in the text and map according to the items in the text.

14.DesignProductionInstruction: here "**GSI MIN CERT** **NO GSI# ENGRAVING ON PCS** **DIRECT SHIP TO CANADA**" is the DesignProductionInstruction for the item similarly find for other items too

15.StampInstruction: here "STAMP 925 LGD ANJ" is the StampInstruction similarly find for other items too

16.OrderGroup: leave it blank

17.Certificate: leave it blank

18.SKUNo: here "LN2095SS-24" is the SKUNo for the item similarly find for other items too

19.Basestoneminwt:leave it blank

20.Basestonemaxwt:leave it blank

21.Basemetalminwt:leave it blank

22.Basemetalmaxwt:leave it blank

23.Productiondeliverydate:leave it blank

24.Expecteddeliverydate:leave it blank

25. create a blank column:leave it blank

26.SetPrice:leave it blank

27.StoneQuality:leave it blank



In [ ]:
import re
import pandas as pd
import pdfplumber
from typing import List, Dict, Optional

# -----------------------------
# Helpers to extract and parse
# -----------------------------

def read_pdf_text(pdf_path: str) -> str:
	full_text = ""
	with pdfplumber.open(pdf_path) as pdf:
		for page in pdf.pages:
			page_text = page.extract_text() or ""
			full_text += page_text + "\n"
	return full_text


def extract_po_number(full_text: str) -> Optional[str]:
	m = re.search(r"PO\s*#\s*:\s*(\d+)", full_text, flags=re.IGNORECASE)
	return m.group(1) if m else None


def split_items(full_text: str) -> List[str]:
	# Find start of the table/listing header; items begin after header where first line like "1." appears
	lines = [ln.strip() for ln in full_text.splitlines()]
	items: List[str] = []
	current: List[str] = []
	inside_items = False
	for ln in lines:
		if re.match(r"^\d+\.", ln):
			inside_items = True
			# if we were collecting, push previous
			if current:
				items.append("\n".join(current).strip())
			current = [ln]
			continue
		if inside_items:
			current.append(ln)
			# Stop condition per spec: item lasts till "**FOR SHIMAYRA**" is seen
			if "**FOR SHIMAYRA**" in ln or "FOR SHIMAYRA" in ln:
				items.append("\n".join(current).strip())
				current = []
				inside_items = False
	# flush trailing (if any)
	if current:
		items.append("\n".join(current).strip())
	return [it for it in items if it]


def parse_first_line_tokens(line: str) -> List[str]:
	# Tokenize keeping slashes and hyphens inside tokens
	return re.findall(r"[A-Za-z0-9/*]+(?:-[A-Za-z0-9]+)*", line)


def find_item_size_and_qty(line: str) -> (Optional[str], Optional[str]):
	# ItemSize: first decimal number in Sr.No. line (e.g., 18.00) or integer with .00
	# OrderQty: number immediately after the ItemSize
	# We'll scan tokens to locate first decimal-like number
	tokens = re.findall(r"\d+\.\d+|\d+", line)
	if not tokens:
		return None, None
	# Prefer decimal as size
	size_idx = None
	for i, tok in enumerate(tokens):
		if re.match(r"^\d+\.\d+$", tok):
			size_idx = i
			break
	# If no decimal present, fallback to first number as size
	if size_idx is None:
		size_idx = 0
	size_val = tokens[size_idx]
	qty_val = None
	if size_idx + 1 < len(tokens):
		qty_val = tokens[size_idx + 1]
	return size_val, qty_val


def find_item_ref_no(line: str) -> Optional[str]:
	m = re.search(r"(\d{5,}/\d+)", line)
	return m.group(1) if m else None


def find_style_code(line: str) -> Optional[str]:
	# StyleCode: second last alphanumeric code in Sr.No. line per example (e.g., ZNLSN00814)
	# Heuristic: consider tokens that are alnum and at least 6 chars with letters and digits
	tokens = parse_first_line_tokens(line)
	candidate_codes: List[str] = []
	for t in tokens:
		if any(c.isalpha() for c in t) and any(c.isdigit() for c in t) and len(t) >= 6:
			candidate_codes.append(t)
	if len(candidate_codes) >= 2:
		return candidate_codes[-2]
	return candidate_codes[-1] if candidate_codes else None


def find_sku(line: str) -> Optional[str]:
	# SKUNo often looks like LN2095SS-24 or similar with hyphen and trailing digits
	m = re.search(r"\b([A-Z]{2,}[A-Z0-9]*-\d{1,})\b", line)
	if m:
		return m.group(1)
	# fallback: third token that contains hyphen
	tokens = parse_first_line_tokens(line)
	for t in tokens:
		if '-' in t and any(ch.isdigit() for ch in t) and any(ch.isalpha() for ch in t):
			return t
	return None


def find_metal_and_tone(block_text: str) -> (Optional[str], Optional[str]):
	# Metal mapping rules from spec
	metal_code = None
	tone = None
	# Look in entire item block
	if re.search(r"\bSILV\b", block_text, flags=re.IGNORECASE):
		metal_code = "AG925"
		# tone: token after SILV
		m = re.search(r"\bSILV\b\s+([A-Z]{1,3})\b", block_text)
		if m:
			tone = m.group(1)
	# 14K gold samples
	if metal_code is None:
		if re.search(r"\bG14\w*\b|\b14K\b|\bGOLD\b", block_text, flags=re.IGNORECASE):
			metal_code = "G14"
			m2 = re.search(r"\bG14\w*\b\s+([A-Z]{1,3})\b", block_text)
			if m2:
				tone = m2.group(1)
	return metal_code, tone


def find_customer_instruction_from_line(line: str) -> Optional[str]:
	# Extract words after SKU until metal keyword (SILV or GOLD/G14/14K) appears
	toks = line.split()
	if not toks:
		return None
	# find index of SKU-like token
	sku_idx = None
	for i, t in enumerate(toks):
		if re.match(r"[A-Z]{2,}[A-Z0-9]*-\d{1,}$", t):
			sku_idx = i
			break
	if sku_idx is None:
		return None
	# collect until metal word
	desc_terms: List[str] = []
	for j in range(sku_idx + 1, len(toks)):
		t = toks[j]
		if t.upper() in {"SILV", "G14", "G14W", "14K", "GOLD"}:
			break
		desc_terms.append(t)
	return " ".join(desc_terms) if desc_terms else None


def extract_design_instructions(block_text: str) -> Optional[str]:
	# Gather phrases inside ** ** within the block, join with space
	phrases = re.findall(r"\*\*([^*]+)\*\*", block_text)
	return " ".join(p.strip() for p in phrases) if phrases else None


def extract_stamp_instruction(block_text: str) -> Optional[str]:
	m = re.search(r"Special\s+Inst\.[^\n]*?STAMP\s+([^,\n]+)", block_text, flags=re.IGNORECASE)
	return m.group(1).strip() if m else None


def extract_special_remarks(block_text: str) -> Optional[str]:
	# Use Special Inst. line as remarks when available
	for ln in block_text.splitlines():
		if ln.upper().startswith("SPECIAL INST."):
			return ln.split("Special Inst.", 1)[-1].strip()
	return None


def parse_items(full_text: str) -> List[Dict[str, str]]:
	item_po_no = extract_po_number(full_text) or ""
	blocks = split_items(full_text)
	parsed: List[Dict[str, str]] = []
	for blk in blocks:
		# First line is Sr.No. line
		first_line = next((ln for ln in blk.splitlines() if re.match(r"^\d+\.\s", ln)), blk.splitlines()[0] if blk.splitlines() else "")
		# SrNo is first integer followed by a dot
		sr_m = re.match(r"^(\d+)\.", first_line.strip())
		sr_no = sr_m.group(1) if sr_m else ""
		item_size, order_qty = find_item_size_and_qty(first_line)
		item_ref_no = find_item_ref_no(first_line) or ""
		style_code = find_style_code(first_line) or ""
		sku_no = find_sku(first_line) or ""
		metal, tone = find_metal_and_tone(blk)
		cust_instr = find_customer_instruction_from_line(first_line) or ""
		design_instr = extract_design_instructions(blk) or ""
		stamp_instr = extract_stamp_instruction(blk) or ""
		special_remarks = extract_special_remarks(blk) or ""
		parsed.append({
			"SrNo": sr_no,
			"StyleCode": style_code,
			"ItemSize": item_size or "",
			"OrderQty": order_qty or "",
			"OrderItemPcs": "",
			"Metal": metal or "",
			"Tone": tone or "",
			"ItemPoNo": item_po_no,
			"ItemRefNo": item_ref_no,
			"StockType": "",
			"MakeType": "",
			"CustomerProductionInstruction": cust_instr,
			"SpecialRemarks": special_remarks,
			"DesignProductionInstruction": design_instr,
			"StampInstruction": stamp_instr,
			"OrderGroup": "",
			"Certificate": "",
			"SKUNo": sku_no,
			"Basestoneminwt": "",
			"Basestonemaxwt": "",
			"Basemetalminwt": "",
			"Basemetalmaxwt": "",
			"Productiondeliverydate": "",
			"Expecteddeliverydate": "",
			"BlankColumn": "",
			"SetPrice": "",
			"StoneQuality": "",
		})
	return parsed


def build_dataframe(records: List[Dict[str, str]]) -> pd.DataFrame:
	columns = [
		"SrNo",
		"StyleCode",
		"ItemSize",
		"OrderQty",
		"OrderItemPcs",
		"Metal",
		"Tone",
		"ItemPoNo",
		"ItemRefNo",
		"StockType",
		"MakeType",
		"CustomerProductionInstruction",
		"SpecialRemarks",
		"DesignProductionInstruction",
		"StampInstruction",
		"OrderGroup",
		"Certificate",
		"SKUNo",
		"Basestoneminwt",
		"Basestonemaxwt",
		"Basemetalminwt",
		"Basemetalmaxwt",
		"Productiondeliverydate",
		"Expecteddeliverydate",
		"BlankColumn",
		"SetPrice",
		"StoneQuality",
	]
	df = pd.DataFrame(records, columns=columns)
	return df


# -----------------------------
# Execute for the Ambition PO
# -----------------------------

# Choose the same path used earlier (update as needed)
pdf_file_path = r'C:\Users\Pratik Mali\Desktop\tools\data\Ambition Jewels Pvt. Ltd (AJP)\NL PO# 101533 AMBITION (SHIMAYRA)-EDITING.pdf'

print(f"Reading PDF from: {pdf_file_path}")
full_text = read_pdf_text(pdf_file_path)

# Parse items and build dataframe
records = parse_items(full_text)
df_items = build_dataframe(records)

print("\nParsed Items DataFrame:")
print(df_items)

# Save to Excel as requested
output_excel = 'structured_items.xlsx'
df_items.to_excel(output_excel, index=False)
print(f"\nData successfully saved to '{output_excel}'")
